File: sdk/scripts/generated-sdk.ts
export class ExampleClient {
  private baseUrl: string;
  token?: string;

  constructor({ baseUrl = 'http://localhost:3000', token }: { baseUrl?: string; token?: string }) {
    this.baseUrl = baseUrl;
    this.token = token;
  }

  async fetch<T = any>({
    method,
    path,
    query,
    body,
    headers: customHeaders = {},
  }: {
    method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
    path: string;
    query?: Record<string, string | number | boolean | null | undefined>;
    body?: T;
    headers?: Record<string, string>;
  }): Promise<Response> {
    const url = new URL(`${this.baseUrl}${path}`);

    if (query) {
      Object.entries(query).forEach(([key, value]) => {
        if (value != null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    const headers = {
      'Content-Type': 'application/json',
      Authorization: this.token ? `Bearer ${this.token}` : '',
      ...customHeaders,
    };

    const options: RequestInit = {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    };

    return fetch(url.toString(), options);
  }

  async take(): Promise<any> {
    try {
      const response = await this.fetch({
        method: 'GET',
        path: '/',
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('Failed to fetch data', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('An unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }

  async *stream(): AsyncGenerator<StreamResponse, void, unknown> {
    const url = new URL(`${this.baseUrl}/stream`);
    const headers = {
      Authorization: this.token ? `Bearer ${this.token}` : '',
    };

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers,
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => null);
      throw new ExampleError('Failed to fetch stream', {
        status: response.status,
        data: errorData,
      });
    }

    if (!response.body) {
      throw new ExampleError('No response body', { status: response.status });
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    let buffer = '';
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const jsonString = line.slice(6).trim();
          if (jsonString) {
            try {
              const data = JSON.parse(jsonString) as StreamResponse;
              yield data;
            } catch (error) {
              console.error('Failed to parse SSE data:', error);
            }
          }
        }
      }
    }
  }

  async getUserById(id: string): Promise<any> {
    const path = `/users/${id}`;

    try {
      const response = await this.fetch({
        method: 'GET',
        path,
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('Failed to fetch user', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('An unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }

  async createUser(body: CreateUserRequest): Promise<CreateUserResponse> {
    try {
      const response = await this.fetch<CreateUserRequest>({
        method: 'POST',
        path: '/users',
        body,
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('Failed to create user', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json() as Promise<CreateUserResponse>;
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('An unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }

  async getError(): Promise<any> {
    try {
      const response = await this.fetch({
        method: 'GET',
        path: '/error',
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('API call failed', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('Unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }

  async getErrorWithSchema(): Promise<{ message: string }> {
    try {
      const response = await this.fetch({
        method: 'GET',
        path: '/errorWithSchema',
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('Request failed', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('An unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }

  async uploadFile({
    file,
  }: {
    file: string; // Base64 encoded file
  }): Promise<any> {
    const path = "/upload";
    const method = "POST";

    const formData = new FormData();
    const blob = new Blob([Buffer.from(file, "base64")], {
      type: "application/octet-stream",
    });
    formData.append("file", blob);

    const headers = {
      Authorization: this.token ? `Bearer ${this.token}` : "",
    };

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: formData,
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError("File upload failed", {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError("Network error during file upload", {
        status: 0,
        data: null,
      });
    }
  }

  async getOpenApiSchema(): Promise<any> {
    try {
      const response = await this.fetch({
        method: 'GET',
        path: '/openapi',
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => null);
        throw new ExampleError('Failed to fetch OpenAPI schema', {
          status: response.status,
          data: errorData,
        });
      }

      return response.json();
    } catch (error) {
      if (error instanceof ExampleError) {
        throw error;
      }
      throw new ExampleError('An unexpected error occurred', {
        status: 500,
        data: error,
      });
    }
  }
}

export class ExampleError extends Error {
  status: number;
  data: any;

  constructor(
    public error: string,
    { status, data }: { status: number; data?: any },
  ) {
    super(error);
    this.status = status;
    this.data = data;
  }
}

export interface StreamResponse {
  count: number;
  timestamp: number;
}

export interface CreateUserRequest {
  name: string;
  email: string;
  age: number;
}

export interface CreateUserResponse {
  message: string;
  data?: any; // Optional field since the schema allows for additional properties
}

File: experiments/sdk/logs/python/dumb/generated-sdk.py
import json
import aiohttp
import urllib.parse
from typing import Any, AsyncGenerator, Dict, Optional, Union


class ExampleClientAsync:
    def __init__(
        self, base_url: str = "http://localhost:3000", token: Optional[str] = None
    ):
        self.base_url = base_url
        self.token = token

    async def fetch(
        self,
        method: str,
        path: str,
        query: Optional[Dict[str, Union[str, int, bool, None]]] = None,
        body: Any = None,
        headers: Optional[Dict[str, str]] = None,
        form_data: Optional[aiohttp.FormData] = None,
    ) -> Any:
        url = urllib.parse.urljoin(self.base_url, path)

        if query:
            params = []
            for key, value in query.items():
                if value is not None:
                    params.append(f"{key}={urllib.parse.quote(str(value))}")
            if params:
                url = f"{url}?{'&'.join(params)}"

        request_headers = {}
        if self.token:
            request_headers["Authorization"] = f"Bearer {self.token}"
        if headers:
            request_headers.update(headers)

        async with aiohttp.ClientSession() as session:
            async with session.request(
                method=method,
                url=url,
                headers=request_headers,
                json=body if body is not None else None,
                data=form_data if form_data is not None else None,
            ) as response:
                return response

    # GET / - x-fern-sdk-group-name: one, x-fern-sdk-method-name: take
    async def take(self) -> Any:
        """Make a GET request to the root endpoint"""
        try:
            response = await self.fetch(
                method="GET",
                path="/",
            )
            
            if response.status != 200:
                raise ExampleError(
                    error=f"Unexpected status code: {response.status}",
                    status=response.status,
                    data=await response.text(),
                )
                
            return await response.json()
        except aiohttp.ClientError as e:
            raise ExampleError(
                error=f"Network error: {str(e)}",
                status=500,
            ) from e

    # GET /stream
    # Method: GET
    # Tags: example-tag
    async def stream(self) -> AsyncGenerator[Dict[str, float], None]:
        """Returns an async generator for the stream endpoint.
        
        Uses server sent events and yields parsed JSON objects with count and timestamp.
        """
        response = await self.fetch(
            method="GET",
            path="/stream",
            headers={"Accept": "text/event-stream"}
        )
        
        if response.status != 200:
            raise ExampleError(
                f"Unexpected status code: {response.status}",
                status=response.status,
                data=await response.text()
            )
            
        async for event in stream_sse_response(response):
            yield event

    # GET /users/{id} - example-tag
    async def get_user(self, id: str) -> Any:
        """Get user by ID
        
        Args:
            id: User ID to retrieve
            
        Returns:
            Any: The user data
            
        Raises:
            ExampleError: If the API returns an error
        """
        response = await self.fetch(
            method="GET",
            path=f"/users/{id}"
        )
        
        if response.status != 200:
            error_data = await response.json()
            raise ExampleError(
                error=f"Failed to get user: {error_data.get('message', 'Unknown error')}",
                status=response.status,
                data=error_data
            )
            
        return await response.json()

    # POST /users
    # Method: POST
    # Tags: users
    async def create_user(
        self,
        name: str,
        email: str,
        age: float,
    ) -> str:
        """
        Create a new user
        
        Args:
            name: User's full name
            email: User's email address
            age: User's age
            
        Returns:
            Success message from the API
            
        Raises:
            ExampleError: If the API returns an error
        """
        body = {
            "name": name,
            "email": email,
            "age": age,
        }
        
        response = await self.fetch(
            method="POST",
            path="/users",
            body=body,
        )
        
        if response.status != 200:
            try:
                error_data = await response.json()
            except:
                error_data = None
            raise ExampleError(
                error=f"Failed to create user: {response.status}",
                status=response.status,
                data=error_data,
            )
            
        response_data = await response.json()
        return response_data["message"]

    # GET /error
    # Method: GET
    # Tags: example-tag
    async def get_error(self) -> None:
        """Always throws an error for testing error handling"""
        response = await self.fetch("GET", "/error")
        
        if response.status >= 400:
            try:
                error_data = await response.json()
                raise ExampleError(
                    error=error_data.get("error", "Unknown error"),
                    status=response.status,
                    data=error_data
                )
            except (json.JSONDecodeError, aiohttp.ContentTypeError):
                raise ExampleError(
                    error=await response.text(),
                    status=response.status
                )

    # GET /errorWithSchema
    # Method: GET
    # Tags: example-tag
    async def get_error_with_schema(self) -> Dict[str, str]:
        """Always throws an error for testing error handling"""
        response = await self.fetch(
            method="GET",
            path="/errorWithSchema",
        )
        
        if response.status != 200:
            try:
                error_data = await response.json()
            except:
                error_data = None
            raise ExampleError(
                error=f"Request failed with status {response.status}",
                status=response.status,
                data=error_data,
            )
            
        return await response.json()

    # POST /upload
    # Tags: (none)
    async def upload_file(self, file_content: str) -> Any:
        """
        Upload a base64 encoded file
        
        Args:
            file_content: Base64 encoded file content as string
            
        Returns:
            Response from the server
            
        Raises:
            ExampleError: If the request fails
        """
        data = aiohttp.FormData()
        data.add_field('file', file_content)
        
        response = await self.fetch(
            method="POST",
            path="/upload",
            body=None,
            headers={"Content-Type": "multipart/form-data"},
            form_data=data
        )
        
        if response.status != 200:
            error_data = await response.json() if response.content_type == "application/json" else await response.text()
            raise ExampleError(
                f"Upload failed with status {response.status}",
                status=response.status,
                data=error_data
            )
            
        return await response.json() if response.content_type == "application/json" else await response.text()

    # GET /openapi
    # tags: 
    async def get_openapi(self) -> Any:
        """Fetch OpenAPI specification
        
        Returns:
            Any: The OpenAPI specification document
        """
        response = await self.fetch(
            method="GET",
            path="/openapi"
        )
        
        if response.status != 200:
            raise ExampleError(
                error=f"Failed to fetch OpenAPI specification: {response.status}",
                status=response.status,
                data=await response.text()
            )
            
        return await response.json()


class ExampleError(Exception):
    def __init__(self, error: str, status: int, data: Any = None):
        super().__init__(error)
        self.status = status
        self.data = data


# this function should not be removed even if not used
async def stream_sse_response(
    response: aiohttp.ClientResponse,
) -> AsyncGenerator[Any, None]:
    if response.content is None:
        return

    async for line in response.content:
        line = line.decode("utf-8")
        if not line.startswith("data: "):
            continue

        data = line[6:].strip()
        if not data:
            continue

        try:
            event = json.loads(data)
            if event.get("event") == "error":
                raise ExampleError(event["data"], status=500)
            yield event
        except json.JSONDecodeError:
            continue








